src/lib.rs

```diff
@@ -36,6 +36,279 @@
 )]
 // This one is hard to avoid.
 #![allow(clippy::multiple_crate_versions)]
+//xx For now:
+#![allow(clippy::not_unsafe_ptr_arg_deref)]
+#![allow(missing_debug_implementations)]
+#![allow(missing_docs)]
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    dead_code,
+    trivial_casts
+)]
+#[allow(clippy::all)]
+mod psa_crypto_binding {
+    include!(concat!(env!("OUT_DIR"), "/shim_bindings.rs"));
+}
+
+#[allow(dead_code)]
+mod constants;
+pub use constants::*;
+
+struct Global {
+    init_succeeded: bool,
+    // Calls to psa_open_key, psa_generate_key and psa_destroy_key
+    // are not thread safe. We work around this bug with the key_mutex.
+    // Mbed issue: https://github.com/ARMmbed/mbed-crypto/issues/266
+    key_mutex: Mutex<()>,
+}
+
+impl Global {
+    fn new() -> Global {
+        Global {
+            init_succeeded: false,
+            key_mutex: Mutex::new(()),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLOBAL: Mutex<Global> = Mutex::new(Global::new());
+}
+
+fn init() -> bool {
+    if GLOBAL.lock().unwrap().init_succeeded {
+        return true;
+    }
+    let status = unsafe { psa_crypto_binding::psa_crypto_init() };
+    if status != PSA_SUCCESS {
+        return false;
+    }
+    GLOBAL.lock().unwrap().init_succeeded = true;
+    true
+}
+
+macro_rules! wrap_any {
+    ($x:expr) => {
+        if !init() {
+            panic!("Error when initialising PSA Crypto")
+        } else {
+            #[allow(unused_unsafe)]
+            unsafe {
+                $x
+            }
+        }
+    };
+}
+
+macro_rules! wrap_status {
+    ($x:expr) => {
+        if !init() {
+            PSA_ERROR_BAD_STATE
+        } else {
+            unsafe { $x }
+        }
+    };
+}
+
+macro_rules! key_lock {
+    ($x:expr) => {{
+        let mutex = &GLOBAL.lock().unwrap().key_mutex;
+        let _guard = mutex.lock().unwrap();
+        $x
+    }};
+}
```

Comment:
This is worth discussing, but what you've suggested could be misleading if there can only be one instance of this structure. What I've done is quite likely not idiomatic Rust, but perhaps a wrapper for a C library should not be expected to look like idiomatic Rust.

(Turning PSA Crypto's single global space of key IDs into a separate space for each application is Parsec's job, right?)

(The mutex is only there to work around a bug in some versions of Mbed Crypto. It would seem wrong to optimise the Rust wrapper's API for working around a bug that will eventually disappear. I don't know what the plans are for the initialisation requirement. If that were to be removed then there would be no need for the mutex or any static data in the Rust wrapper, I think.)

Reply:
> This is worth discussing, but what you've suggested could be misleading if there can only be one instance of this structure.

Good point! We can explore options how to create singletons. Having one instance of `PsaCrypto` would allow you to leverage the Rust borrow-checker to have thread-safety without a mutex.

> perhaps a wrapper for a C library should not be expected to look like idiomatic Rust

I tend to disagree with that point. I think Rust developpers consuming this crate will expect it to follow the usual Rust patterns. In the end "idiomatic" just means easy to use in a Rust context: safe and convenient abstractions. I do not think we should force the C style on Rust.

> (Turning PSA Crypto's single global space of key IDs into a separate space for each application is Parsec's job, right?)

That is correct! By "abstraction" I do not mean to achieve that. I do think that we should provide the exact same features/functionalities than PSA Crypto with the same semantics. Still use Key ID. By abstraction I mean using the high-level features of Rust: stricter types with methods being the most important one for a Rust to C wrapper in my opinion.

> (The mutex is only there to work around a bug in some versions of Mbed Crypto. It would seem wrong to optimise the Rust wrapper's API for working around a bug that will eventually disappear. I don't know what the plans are for the initialisation requirement. If that were to be removed then there would be no need for the mutex or any static data in the Rust wrapper, I think.)

In this message I meant the mutex that is needed around the `GLOBAL` structure.

Reply:
Seems this hasn't been mentioned yet: [`std::sync::Once`](https://doc.rust-lang.org/std/sync/struct.Once.html); Not necessarily saying we should use it here, it just looks useful/aligned with the task at hand

Also, while I'm not a fan of the "panic if used twice" pattern, [this](https://docs.rust-embedded.org/book/peripherals/singletons.html) page has some good points about having the compiler do some of the sanity checks for you.